lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -396,259 +304,353 @@ impl FwCfg {
             (FW_CFG_IOP_DMA_LO, 4),
         ];
         let this = self.clone();
-        let piofn = Arc::new(move |port: u16, rwo: RWOp| this.pio_rw(port, rwo))
-            as Arc<PioFn>;
+        let piofn = Arc::new(move |port: u16, rwo: RWOp| match rwo {
+            RWOp::Read(ro) => this.pio_read(port, ro),
+            RWOp::Write(wo) => this.pio_write(port, wo),
+        }) as Arc<PioFn>;
         for (port, len) in ports.iter() {
             pio.register(*port, *len, piofn.clone()).unwrap()
         }
     }
 
-    fn pio_rw(&self, port: u16, rwo: RWOp) {
+    /// Change the [RamFb] attachment for any [Entry::RamFb] entry(s)
+    pub fn attach_ramfb(
+        &self,
+        ramfb: Option<Arc<RamFb>>,
+    ) -> Option<Arc<RamFb>> {
+        let mut state = self.state.lock().unwrap();
+        std::mem::replace(&mut state.ramfb, ramfb)
+    }
+
+    pub fn insert_legacy(
+        &self,
+        id: LegacyId,
+        entry: Entry,
+    ) -> Result<(), InsertError> {
+        let mut state = self.state.lock().unwrap();
+        state.directory.insert_legacy(id, entry)
+    }
+    pub fn insert_named(
+        &self,
+        name: &str,
+        entry: Entry,
+    ) -> Result<u16, InsertError> {
+        let mut state = self.state.lock().unwrap();
+        let selector = state
+            .directory
+            .next_named_selector()
+            .ok_or(InsertError::NoCapacity)?;
+        state.directory.insert(selector, name.to_owned(), entry)?;
+        Ok(selector)
+    }
+
+    pub fn remove(&self, selector: u16) -> Option<Entry> {
+        let mut state = self.state.lock().unwrap();
+        let entry = state.directory.remove(selector)?;
+        Self::validate_selected(&mut state);
+        Some(entry)
+    }
+    pub fn remove_named(&self, name: &str) -> Option<Entry> {
+        let mut state = self.state.lock().unwrap();
+        let selector = state.directory.named_selector(name)?;
+        let entry = state.directory.remove(selector).unwrap();
```

Comment:
nit, take it or leave it: maybe this `unwrap()` could be an `expect` with a message noting that we already would have returned `None` if there was no selector for that name? mostly for the benefit of a reader rather than because i expect this to panic.

not a big deal either way though.

Reply:
Done.